File: Easy/0206. Reverse Linked List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseList(ListNode head) {
        // Set it to nothing to - last node points to nothing
        return reverse(head, null);
    }
    
    private ListNode reverse(ListNode head, ListNode prev){
        if(head == null)
            return prev;
        
        // Store next Node
        var next = head.next;
        // Next node should point to previous
        head.next = prev;
        
        // Set previous node to current one
        return reverse(next, head);
    }
    
    public ListNode ReverseList2(ListNode head) {
        // Set it to nothing to - last node points to nothing
        ListNode prevHead = null;
        
        while(head != null){
            // Store next Node
            var nextHead = head.next;
            
            // Next node should point to previous
            head.next = prevHead;
            // Set previous node to current one
            prevHead = head;
            
            // Restore Original next Item
            head = nextHead;
        }
        
        return prevHead;
    }
}

File: Linked List/0206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev, curr = None, head

        while curr:
            nxt = curr.next # save the next node
            curr.next = prev
            prev = curr
            curr = nxt

        return prev

    def reverseList2(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        newHead = head
        if head.next:
            newHead = self.reverseList2(head.next)
            head.next.next = head
        head.next = None

        return newHead